### PR TITLE
[FLINK-1567] Add switch to use the AvroSerializer for GenericTypeInfo

### PR DESCRIPTION
**Pull request by rmetzger**
Note this is a fix for 0.8 only.
The fix for master is contained in #393

### CODE REVIEW

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/GenericTypeInfo.java [hunk 1]
```diff
@@ -66,6 +71,9 @@ public boolean isKeyType() {
 
 	@Override
 	public TypeSerializer<T> createSerializer() {
+		if(useAvroSerializer) {
+			return new AvroSerializer<T>(this.typeClass);
+		}
 		return new KryoSerializer<T>(this.typeClass);
```
Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/GenericTypeInfo.java]: If we enclose the ```return new KryoSerializer``` in an else block, then the control flow looks nicer imho.

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java [hunk 2]
```diff
@@ -91,7 +98,22 @@ public T createInstance() {
...
+			Output output = new Output(baout);
+
+			kryo.writeObject(output, from);
+
+			output.close();
+
+			ByteArrayInputStream bain = new ByteArrayInputStream(baout.toByteArray());
+			Input input = new Input(bain);
+
+			return (T)kryo.readObject(input, from.getClass());
```
Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java]: Which cases do we cover with this method?

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java [hunk 3]
```diff
@@ -154,6 +176,7 @@ private void checkAvroInitialized() {
 	private void checkKryoInitialized() {
 		if (this.kryo == null) {
 			this.kryo = new Kryo();
+			this.kryo.register(GenericData.Array.class, new KryoSerializer.SpecificInstanceCollectionSerializer(ArrayList.class));
```
Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java]: Without that, it does not work?

**Comment by tillrohrmann** [comment]
LGTM besides of my small comments.

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java (hunk 3 above)
Comment (rmetzger) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java]: Without that, Kryo's `CollectionSerializer` is not able to properly re-instantiate the `GenericData.Array`.
It fails with the following error:
```
Class cannot be created (missing no-arg constructor): org.apache.avro.generic.GenericData$Array
Serialization trace:
type_array_boolean (org.apache.flink.api.io.avro.generated.User)
com.esotericsoftware.kryo.KryoException: Class cannot be created (missing no-arg constructor): org.apache.avro.generic.GenericData$Array
Serialization trace:
type_array_boolean (org.apache.flink.api.io.avro.generated.User)
	at com.esotericsoftware.kryo.Kryo$DefaultInstantiatorStrategy.newInstantiatorOf(Kryo.java:1228)
	at com.esotericsoftware.kryo.Kryo.newInstantiator(Kryo.java:1049)
	at com.esotericsoftware.kryo.Kryo.newInstance(Kryo.java:1058)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.createCopy(CollectionSerializer.java:124)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.copy(CollectionSerializer.java:128)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.copy(CollectionSerializer.java:22)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.AsmCacheFields$AsmObjectField.copy(AsmCacheFields.java:177)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:634)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at org.apache.flink.api.java.typeutils.runtime.AvroSerializer.copy(AvroSerializer.java:102)
	at org.apache.flink.api.common.operators.GenericDataSourceBase.executeOnCollections(GenericDataSourceBase.java:196)
	at org.apache.flink.api.common.operators.CollectionExecutor.executeDataSource(CollectionExecutor.java:170)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:132)
	at org.apache.flink.api.common.operators.CollectionExecutor.executeUnaryOperator(CollectionExecutor.java:180)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:126)
	at org.apache.flink.api.common.operators.CollectionExecutor.executeUnaryOperator(CollectionExecutor.java:180)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:126)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:108)
	at org.apache.flink.api.common.operators.CollectionExecutor.executeDataSink(CollectionExecutor.java:159)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:135)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:108)
	at org.apache.flink.api.common.operators.CollectionExecutor.execute(CollectionExecutor.java:99)
	at org.apache.flink.api.java.CollectionEnvironment.execute(CollectionEnvironment.java:34)
	at org.apache.flink.test.util.JavaProgramTestBase$CollectionTestEnvironment.execute(JavaProgramTestBase.java:269)
	at org.apache.flink.api.io.avro.AvroPojoTest.runProgram(AvroPojoTest.java:195)
	at org.apache.flink.api.io.avro.AvroPojoTest.testProgram(AvroPojoTest.java:115)
	at org.apache.flink.test.util.JavaProgramTestBase.testJobCollectionExecution(JavaProgramTestBase.java:172)
```

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java (hunk 2 above)
Comment (rmetzger) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/AvroSerializer.java]: Some Kryo serializers do not implement the copy method. In that case, we do a serialization copy.

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/GenericTypeInfo.java (hunk 1 above)
Comment (rmetzger) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/GenericTypeInfo.java]: I respect [yho](http://www.urbandictionary.com/define.php?term=Yho&defid=5440750) and update the PR.

**Comment by rmetzger** [comment]
I've updated the PR addressing Till's comments and adding a `GenericAvroTypeInfo` giving users an additional way to enforce the Avro serializer.

**Comment by StephanEwen** [comment]
I assume that this is targeted at the 0.8 branch?

For the 0.9 branch, it would be nice to use the execution config instead to define which serializer to instantiate.

**Comment by rmetzger** [comment]
Exactly, the change is only for 0.8.
For 0.9 its already merged to master using the ExecutionConfig.

**Comment by StephanEwen** [comment]
Looks good to merge

**Comment by rmetzger** [comment]
Thank you. I'm going to merge it.

**Pull request merged**